debug/debugger.go

```diff
@@ -0,0 +1,859 @@
+}
+
+type printHook struct {
+	prnt.Hook
+	d *debugger
+}
+
+func (h *printHook) Print(_ prnt.Context, str string) error {
+	if h == nil || h.d == nil {
+		return nil
+	}
+	h.d.eventHandler(Event{Type: StdoutEventType, Message: str})
+	return nil
+}
+
+type DebuggerOption func(*debugger)
+
+func NewDebugger(options ...DebuggerOption) Debugger {
+	return newDebugger(options...)
+}
+
+func newDebugger(options ...DebuggerOption) *debugger {
+	d := &debugger{
+		eventHandler: newNopEventHandler(),
+		logger:       logging.NewNoOpLogger(),
+	}
+	d.printHook = &printHook{d: d}
+
+	for _, option := range options {
+		option(d)
+	}
+
+	return d
+}
+
+func SetLogger(logger logging.Logger) DebuggerOption {
+	return func(d *debugger) {
+		d.logger = logger
+	}
+}
+
+func SetEventHandler(handler EventHandler) DebuggerOption {
+	return func(d *debugger) {
+		d.eventHandler = handler
+	}
+}
+
+type LaunchEvalProperties struct {
+	LaunchProperties
+	Query     string
+	Input     interface{}
+	InputPath string
+}
+
+type LaunchTestProperties struct {
+	LaunchProperties
+	Run string `json:"run"`
+}
+
+type LaunchProperties struct {
+	BundlePaths  []string
+	DataPaths    []string
+	StopOnResult bool
+	StopOnEntry  bool
+	StopOnFail   bool
+	EnablePrint  bool
+	SkipOps      []topdown.Op
+	RuleIndexing bool
+}
+
+func (lp LaunchProperties) String() string {
+	b, err := json.Marshal(lp)
+	if err != nil {
+		return "{}"
+	}
+	return string(b)
+}
+
+func (d *debugger) LaunchEval(ctx context.Context, props LaunchEvalProperties) (Session, error) {
+	store := inmem.New()
+	txn, err := store.NewTransaction(ctx, storage.TransactionParams{Write: true})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create store transaction: %v", err)
+	}
+
+	regoArgs := []func(*rego.Rego){
+		rego.Query(props.Query),
+		rego.Store(store),
+		rego.Transaction(txn),
+	}
+
+	if props.SkipOps == nil {
+		props.SkipOps = []topdown.Op{topdown.IndexOp, topdown.RedoOp, topdown.SaveOp, topdown.UnifyOp}
+	}
+
+	if props.EnablePrint {
+		regoArgs = append(regoArgs, rego.EnablePrintStatements(true),
+			rego.PrintHook(d.printHook))
+	}
+
+	if len(props.DataPaths) > 0 {
+		regoArgs = append(regoArgs, rego.Load(props.DataPaths, nil))
+	}
+
+	for _, bundlePath := range props.BundlePaths {
+		regoArgs = append(regoArgs, rego.LoadBundle(bundlePath))
+	}
+
+	if props.InputPath != "" && props.Input != nil {
+		return nil, fmt.Errorf("cannot specify both input and input path")
+	}
+
+	if props.Input != nil {
+		regoArgs = append(regoArgs, rego.Input(props.Input))
+	} else if props.InputPath != "" {
+		input, err := readInput(props.InputPath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read input: %v", err)
+		}
+		regoArgs = append(regoArgs, rego.Input(input))
+	}
+
+	r := rego.New(regoArgs...)
+
+	pq, err := r.PrepareForEval(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to prepare query for evaluation: %v", err)
+	}
+
+	if err := store.Commit(ctx, txn); err != nil {
```

Comment:
We're committing any data added during `PrepareForEval`. I found that if we don't do this, then we won't be able to read this data from the store in `thread.dataVars()`. This might not be the correct place or method to do this, though; so if you can point me to a more proper solution, I'm all ears 🙂.

Reply:
Thanks for the context. Maybe add a comment mentioning the same and we can improve later if needed.